Movie Search/SearchController.swift

```diff
@@ -0,0 +1,122 @@
+//
+//  SearchController.swift
+//  Movie Search
+//
+//  Created by Alexey Grabik on 15/04/2020.
+//
+
+import UIKit
+
+enum SearchStatus {
+    case error(String), nothingRequested, noResult, success
+}
+
+class SearchController: NSObject {
+    
+    private let networkService: NetworkService
+    
+    private var moviesModels = [MovieModel]() {
+        didSet {
+            downloadQueue.cancelAllOperations()
+        }
+    }
+    private let downloadQueue = OperationQueue()
+    
+    var searchPerformedCallback: ((SearchStatus) -> Void)?
+    
+    init(networkService: NetworkService = NetworkAPI()) {
+        self.networkService = networkService
+        self.downloadQueue.maxConcurrentOperationCount = Constants.maxConcurrentOperationCount
+        super.init()
+    }
+    
+    func search(for movieTitle: String) {
+        NSObject.cancelPreviousPerformRequests(withTarget: self) 
+        perform(#selector(self.performSearch(_:)),
+                with: movieTitle,
+                afterDelay: Constants.searchThrottleDelay)
+    }
+}
+
+private extension SearchController {
+    
+    class func status(with networkError: NetworkServiceError?) -> SearchStatus? {
+        guard let error = networkError else { return nil }
+        switch error {
+        case .badResponse:
+            return .error("Bad response")
+        case .parsingError:
+            return .error("Parsing error")
+        case .constructingRequestError:
+            return .error("Constructing request error")
+        }
+    }
+    
+    @objc
+    func performSearch(_ movieTitle: String) {
+        if movieTitle.count > 0 {
+            networkService.search(for: movieTitle) { moviesModels, error in
+                self.moviesModels = moviesModels ?? []
+                if let status = Self.status(with: error) {
+                    self.searchPerformedCallback?(status)
+                }
+                else if self.moviesModels.isEmpty {
+                    self.searchPerformedCallback?(.noResult)
+                }
+                else {
+                    self.searchPerformedCallback?(.success)
+                }
+            }
+        }
+        else {
+            moviesModels = []
+            searchPerformedCallback?(.nothingRequested)
+        }
+    }
+    
+    func tryToLoadPhoto(for indexPath: IndexPath, in tableView: UITableView) {
+        let movieModel = moviesModels[indexPath.row]
```

Comment:
this doesn't seem safe - check whether calling this will not cause an out of bounds exception before subscripting it.